robot/impl/resource_manager.go

```diff
@@ -819,50 +819,63 @@ func (manager *resourceManager) completeConfig(
 }
 
 func (manager *resourceManager) completeConfigForRemotes(ctx context.Context, lr *localRobot) {
+	// Add remotes in parallel. This is particularly useful in cases where
+	// there are many remotes that are offline or slow to start up.
+	var remoteErrGroup errgroup.Group
+	remoteErrGroup.SetLimit(5)
 	for _, resName := range manager.resources.FindNodesByAPI(client.RemoteAPI) {
 		gNode, ok := manager.resources.Node(resName)
 		if !ok || !gNode.NeedsReconfigure() {
-			continue
-		}
-		var verb string
-		if gNode.IsUninitialized() {
-			verb = "configuring"
-		} else {
-			verb = "reconfiguring"
+			return
 		}
-		manager.logger.CInfow(ctx, fmt.Sprintf("Now %s a remote", verb), "resource", resName)
-		switch resName.API {
-		case client.RemoteAPI:
-			remConf, err := resource.NativeConfig[*config.Remote](gNode.Config())
-			if err != nil {
-				manager.logger.CErrorw(ctx, "remote config error", "error", err)
-				continue
-			}
+		processRemote := func() {
+			var verb string
 			if gNode.IsUninitialized() {
-				gNode.InitializeLogger(
-					manager.logger, fromRemoteNameToRemoteNodeName(remConf.Name).String(),
-				)
-			}
-			// this is done in config validation but partial start rules require us to check again
-			if _, err := remConf.Validate(""); err != nil {
-				gNode.LogAndSetLastError(
-					fmt.Errorf("remote config validation error: %w", err), "remote", remConf.Name)
-				continue
+				verb = "configuring"
+			} else {
+				verb = "reconfiguring"
 			}
-			rr, err := manager.processRemote(ctx, *remConf, gNode)
-			if err != nil {
-				gNode.LogAndSetLastError(
-					fmt.Errorf("error connecting to remote: %w", err), "remote", remConf.Name)
-				continue
+			manager.logger.CInfow(ctx, fmt.Sprintf("Now %s a remote", verb), "resource", resName)
+			switch resName.API {
+			case client.RemoteAPI:
+				remConf, err := resource.NativeConfig[*config.Remote](gNode.Config())
+				if err != nil {
+					manager.logger.CErrorw(ctx, "remote config error", "error", err)
+					return
+				}
+				if gNode.IsUninitialized() {
+					gNode.InitializeLogger(
+						manager.logger, fromRemoteNameToRemoteNodeName(remConf.Name).String(),
+					)
+				}
+				// this is done in config validation but partial start rules require us to check again
+				if _, err := remConf.Validate(""); err != nil {
+					gNode.LogAndSetLastError(
+						fmt.Errorf("remote config validation error: %w", err), "remote", remConf.Name)
+					return
+				}
+				rr, err := manager.processRemote(ctx, *remConf, gNode)
```

Comment:
I know you looked at a lot of this code re: locks. What does `processRemote` and the following `addRemote` do again? I'm assuming one of them is getting the remote resources (via the `ResourceNames` API call) and creating resource graph nodes for them?

I imagine any lock acquisitions we're doing here are simply about protecting access to maps? And not necessarily about ensuring some kind of logical consistency between resources from different remotes?